gcs_oauth2_boto_plugin/oauth2_helper.py

```diff
@@ -78,29 +79,37 @@ def OAuth2ClientFromBotoConfig(
   if cred_type == oauth2_client.CredTypes.OAUTH2_SERVICE_ACCOUNT:
     service_client_id = config.get('Credentials', 'gs_service_client_id', '')
     private_key_filename = config.get('Credentials', 'gs_service_key_file', '')
-    with io.open(private_key_filename, 'r', encoding='utf-8') as private_key_file:
+    with io.open(private_key_filename, 'rb') as private_key_file:
       private_key = private_key_file.read()
 
-    json_key_dict = None
+    keyfile_is_utf8 = False
     try:
-      json_key_dict = json.loads(private_key)
-    except ValueError:
+      private_key = private_key.decode(UTF8)
+      # P12 keys won't be encoded as UTF8 bytes.
+      keyfile_is_utf8 = True
+    except UnicodeDecodeError:
       pass
-    if json_key_dict:
-      for json_entry in ('client_id', 'client_email', 'private_key_id',
-                         'private_key'):
-        if json_entry not in json_key_dict:
-          raise Exception('The JSON private key file at %s '
-                          'did not contain the required entry: %s' %
-                          (private_key_filename, json_entry))
-
-      return oauth2_client.OAuth2JsonServiceAccountClient(
-          json_key_dict, access_token_cache=token_cache,
-          auth_uri=provider_authorization_uri, token_uri=provider_token_uri,
-          disable_ssl_certificate_validation=not(config.getbool(
-              'Boto', 'https_validate_certificates', True)),
-          proxy_host=proxy_host, proxy_port=proxy_port,
-          proxy_user=proxy_user, proxy_pass=proxy_pass)
+
+    if keyfile_is_utf8:
+      json_key_dict = None
+      try:
+        json_key_dict = json.loads(private_key)
+      except ValueError:
+        pass
```

Comment:
We might want to throw an exception here -- the way the code path has been refactored, this `pass` was okay because trying to construct the p12 credential would fail -- here, we'd just skip the `if json_key_dict:` block and continue on.

Can we make this throw an exception, saying that we couldn't parse the JSON keyfile at `<path>`?

Reply:
That is, I think these lines should look like:

```
    if keyfile_is_utf8:
      try:
        json_key_dict = json.loads(private_key)
      except ValueError:
        raise Exception(
            'Could not parse JSON keyfile "%s" as valid JSON' %
            private_key_filename)
      for json_entry in (...
```

Reply:
Sounds reasonable to me! Adding that now 🙂